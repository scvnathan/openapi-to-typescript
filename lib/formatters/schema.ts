import { JSONSchema } from 'json-schema-ref-parser'
import { compileSchema } from '../compile'
import { Formatter } from '../formatter'
export class SchemaFormatter extends Formatter<JSONSchema> {
  public readonly name: string

  constructor(name: string) {
    super()
    this.name = name
  }

  public async render(schema: JSONSchema) {
    return await compileSchema(schema, this.name)
  }
}
