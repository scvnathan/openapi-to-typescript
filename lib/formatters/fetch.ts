
import { Operation } from 'lib/operation'
import { get } from 'lodash'
import { Formatter } from '../formatter'
import { OpenAPIObject, ParameterLocation } from '../typings/openapi'
import { type } from 'os';

const withoutIndentation = (s: string) => s.replace(/(\s*[\n]+)\s*/g, '\n').trim()
const withoutEmptyLines = (s: string) => s.replace(/\n+/g, '\n')

export class FetchClientFormatter extends Formatter<Operation> {
  public url: string
  constructor({url}: any = {}) {
    super()
    this.url = url
  }

  /* tslint:disable:max-line-length */
  public async renderBoilerplate(apiSchema?: OpenAPIObject) {
    const url = this.url || get(apiSchema, 'servers[0].url')
    return withoutIndentation(`
      const fetch = require('node-fetch')
      const pick = (obj:any, keys:string[]) => keys.reduce((picked, key) => obj[key] !== undefined ? Object.assign(picked, {[key]: obj[key]}) : picked, {})
      const substitutePath = (pathName:string, body:any) => pathName.split('/').map(dir => dir.startsWith('{') ? body[dir.slice(1,-1)] : dir).join('/')
      const encodeQuery = (obj:any, keys:string[]) => require('querystring').encode(pick(obj, keys))
      const API_URL = ${JSON.stringify(url)}
    `)
  }
  /* tslint:enable:max-line-length */

  public async render(operation: Operation) {
    const { operationName, requestTypeName, responseTypeName } = this.names(operation)
    const typedef = `export type ${operationName} = (payload: ${requestTypeName}) => Promise<${responseTypeName}>;`
    const fetchWrapper = withoutEmptyLines(`const ${operationName} =
      async (body:${requestTypeName}, options:any):Promise<${responseTypeName}> => {
        return fetch(${this.urlSnippet(operation)}, {
          ...options,
          method: ${JSON.stringify(operation.method)},
          ${operation.operationObject.requestBody ? 'body: JSON.stringify(body),' : ''}
        }).then((res:any) => res.json())
      }`)
    return `${typedef}\n${fetchWrapper}`
  }

  public names(operation: Operation) {
    const operationName = operation.name
    const operationTypeName = operation.name + 'OperationType'
    const requestTypeName = operationName + 'Request'
    const responseTypeName = operationName + 'Result'
    return { operationName, operationTypeName, requestTypeName, responseTypeName }
  }

  private urlSnippet(operation: Operation) {
    const queryParameterNames = operation.parameterNamesIn('query')
    const endpointURL = operation.hasAnyParametersIn('path') ?
      `API_URL + substitutePath(${JSON.stringify(operation.pathName)}, body)` :
      'API_URL'
    return queryParameterNames.length ?
      `[${endpointURL}, encodeQuery(body, ${JSON.stringify(queryParameterNames)})].filter(x=>x).join('?')` :
      endpointURL
  }
}
