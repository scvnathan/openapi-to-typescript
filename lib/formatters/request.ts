import { get } from 'lodash'

import { JSONSchema } from 'json-schema-ref-parser'
import { getSchemaName, getSchemaNameByRef } from '../compile'
import { Formatter } from '../formatter'
import { Operation } from '../operation'
import { ParameterObject } from '../typings/openapi'

export class RequestTypeFormatter extends Formatter<Operation> {
  public async render(operation: Operation) {
    const typeName = `${operation.name}Request`
    const parameters = operation.operationObject.parameters || []
    return await this.toTypescriptInterface(operation, typeName, parameters as ParameterObject[])
  }

  public async toTypescriptInterface(operation: Operation, typeName: string, parameters: ParameterObject[]) {
    const requestSchema = get(operation.operationObject, 'requestBody.content["application/json"].schema')
    const aliasedType = requestSchema && requestSchema.$ref && getSchemaNameByRef(requestSchema.$ref)
    if (aliasedType) return `export type ${typeName} = ${aliasedType}`
    return [
      `export interface ${typeName} {`,
        ...parameters.map((param) =>
          `  ${param.name}: ${getSchemaName(param.schema as JSONSchema, param.name)}`),
      '}',
    ].join('\n')
  }
}
