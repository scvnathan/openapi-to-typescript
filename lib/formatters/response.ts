import { JSONSchema } from 'json-schema-ref-parser'
import { camelCase, get, upperFirst } from 'lodash'

import { Operation } from 'lib/operation'
import { compileSchema } from '../compile'
import { Formatter } from '../formatter'

export class ResultTypeFormatter extends Formatter<Operation> {
  public readonly contentType: string = 'application/json'

  public async render(operation: Operation) {
    const definitions = this.getResponseSchemaDefinitions(operation)
    const compilations = await Promise.all(
      definitions.map(({ schema, typeName }) =>
        this.compileDefinition(schema, typeName)),
    )
    return compilations.join('\n')
  }

  private async compileDefinition(schema: JSONSchema, typeName: string) {
    return schema ?
      compileSchema(schema, typeName) :
      `export type ${typeName} = any`
  }

  private getResponseSchemaDefinitions(operation: Operation): ResponseSchemaDefinition[] {
    const responsesByStatusCode = get(operation.operationObject, 'responses', {})
    const statusCodes = this.getStatusCodes(responsesByStatusCode)
    return statusCodes.map(((statusCode, index): ResponseSchemaDefinition => {
      const key = `responses['${statusCode}'].content[${this.contentType}].schema`
      const typeName = this.typeNameFor(operation, index === 0 ? 'Result' : statusCode)
      const schema = get(operation.operationObject, key)
      return { schema, statusCode, typeName }
    }))
  }

  private getStatusCodes(responsesByStatusCode: any) {
    const whitelist: (number|string)[] = ['default']
    return Object.keys(responsesByStatusCode)
      .map((n) => whitelist.includes(n) ? n : parseInt(n, 10))
      .filter((n) => whitelist.includes(n) || n >= 200 && n < 400)
      .sort()
  }

  private typeNameFor(operation: Operation, suffix: number|string) {
    return upperFirst(camelCase(`
      ${operation.name}
      ${suffix === 'default' ? 'fallback' : suffix.toString()}
    `))
  }
}

interface ResponseSchemaDefinition {
  schema: JSONSchema,
  typeName: string,
  statusCode: number|string|'default'
}
