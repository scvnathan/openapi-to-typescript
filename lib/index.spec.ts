import execa from 'execa'
import { FetchClientFormatter } from './formatters'
import { GenerateTypings, GenerateTypingsOptions } from './index'
import { OpenAPIObject } from './typings/openapi'

describe('GenerateTypings', () => {
  describe('given an empty openapi schema', () => {
    const schema: OpenAPIObject = {
      openapi: '3.0.0',
      info: {
        version: '1.0.0',
        title: 'Some API',
      },
      components: {
        schemas: {},
      },
      paths: {},
    }
    itShouldGenerateValidTypingsFromSchema(schema)
    it('should respond nothing', async () => {
      const generated = await GenerateTypings(schema)
      expect(generated).toMatch(/^\s*$/)
      expect(generated).toEqual('')
    })
  })

  describe('given the Petstore schema', () => {
    const schema = require('../fixtures/petstore.json')
    itShouldGenerateValidTypingsFromSchema(schema)
    it('should output types for components.schemas', async () => {
      const generated = await GenerateTypings(schema)
      expect(generated).toContain('export interface Pet')
      expect(generated).toContain('export type Pets = (Pet)[]')
    })
  })

  describe('given the Petstore Expanded schema', () => {
    const schema = require('../fixtures/petstoreExpanded.json')
    itShouldGenerateValidTypingsFromSchema(schema)
    it('should output types for components.schemas', async () => {
      const generated = await GenerateTypings(schema)
      expect(generated).toContain('export type Pet = NewPet & {')
      expect(generated).toContain('NewPet ')
    })
    it('should output types for paths', async () => {
      const generated = await GenerateTypings(schema)
      expect(generated).toContain('FindPetByIdRequest')
      expect(generated).toContain('FindPetByIdResult') 
      expect(generated).toContain('export type FindPetByIdResult = Pet;\n')
      expect(generated).toContain('FindPetByIdFallback')
      expect(generated).toContain('export type FindPetByIdFallback = Error;\n')
    })

    describe('FetchClientFormatter', () => {
      const options = {
        operationFormatters: [new FetchClientFormatter()],
      }
      itShouldGenerateValidTypingsFromSchema(schema, options)

      it('should generate client actions', async () => {
        const generated = await GenerateTypings(schema, options)
        expect(generated).toContain('method: "post"')
      })
    })
  })
})

function itShouldGenerateValidTypingsFromSchema(schema: any, options?: GenerateTypingsOptions) {
  it('should match generated snapshot', async () => {
    expect((await GenerateTypings(schema, options))).toMatchSnapshot()
  })

  it('does not contain $magic$', async () => {
    expect(await GenerateTypings(schema, options)).not.toContain('$magic$')
  })

  it('should evaluate', async () => {
    const generated = await GenerateTypings(schema, options)
    if (/\S+/.test(generated)) await execa.stdout('ts-node', ['--eval', generated])
  })
}
