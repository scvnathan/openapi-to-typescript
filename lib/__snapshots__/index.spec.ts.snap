// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateTypings given an empty openapi schema should match generated snapshot 1`] = `""`;

exports[`GenerateTypings given the Petstore Expanded schema FetchClientFormatter should match generated snapshot 1`] = `
"export type Pet = NewPet & {
  id: number;
  [k: string]: any;
};

export interface NewPet {
  name: string;
  tag?: string;
  [k: string]: any;
}

export interface Error {
  code: number;
  message: string;
  [k: string]: any;
}

const fetch = require('node-fetch')
const pick = (obj:any, keys:string[]) => keys.reduce((picked, key) => obj[key] !== undefined ? Object.assign(picked, {[key]: obj[key]}) : picked, {})
const substitutePath = (pathName:string, body:any) => pathName.split('/').map(dir => dir.startsWith('{') ? body[dir.slice(1,-1)] : dir).join('/')
const encodeQuery = (obj:any, keys:string[]) => require('querystring').encode(pick(obj, keys))
const API_URL = \\"http://petstore.swagger.io/api\\"
export interface FindPetsRequest {
  tags: string[]
  limit: number
}
export type FindPetsResult = (Pet)[];

export type FindPetsFallback = Error;

export type FindPets = (payload: FindPetsRequest) => Promise<FindPetsResult>;
const FindPets =
      async (body:FindPetsRequest, options:any):Promise<FindPetsResult> => {
        return fetch([API_URL, encodeQuery(body, [\\"tags\\",\\"limit\\"])].filter(x=>x).join('?'), {
          ...options,
          method: \\"get\\",
          
        }).then((res:any) => res.json())
      }
export type AddPetRequest = NewPet
export type AddPetResult = Pet;

export type AddPetFallback = Error;

export type AddPet = (payload: AddPetRequest) => Promise<AddPetResult>;
const AddPet =
      async (body:AddPetRequest, options:any):Promise<AddPetResult> => {
        return fetch(API_URL, {
          ...options,
          method: \\"post\\",
          body: JSON.stringify(body),
        }).then((res:any) => res.json())
      }
export interface FindPetByIdRequest {
  id: number
}
export type FindPetByIdResult = Pet;

export type FindPetByIdFallback = Error;

export type FindPetById = (payload: FindPetByIdRequest) => Promise<FindPetByIdResult>;
const FindPetById =
      async (body:FindPetByIdRequest, options:any):Promise<FindPetByIdResult> => {
        return fetch(API_URL + substitutePath(\\"/pets/{id}\\", body), {
          ...options,
          method: \\"get\\",
          
        }).then((res:any) => res.json())
      }
export interface DeletePetRequest {
  id: number
}
export type DeletePetResult = any
export type DeletePetFallback = Error;

export type DeletePet = (payload: DeletePetRequest) => Promise<DeletePetResult>;
const DeletePet =
      async (body:DeletePetRequest, options:any):Promise<DeletePetResult> => {
        return fetch(API_URL + substitutePath(\\"/pets/{id}\\", body), {
          ...options,
          method: \\"delete\\",
          
        }).then((res:any) => res.json())
      }"
`;

exports[`GenerateTypings given the Petstore Expanded schema should match generated snapshot 1`] = `
"export type Pet = NewPet & {
  id: number;
  [k: string]: any;
};

export interface NewPet {
  name: string;
  tag?: string;
  [k: string]: any;
}

export interface Error {
  code: number;
  message: string;
  [k: string]: any;
}

export interface FindPetsRequest {
  tags: string[]
  limit: number
}
export type FindPetsResult = (Pet)[];

export type FindPetsFallback = Error;

export type AddPetRequest = NewPet
export type AddPetResult = Pet;

export type AddPetFallback = Error;

export interface FindPetByIdRequest {
  id: number
}
export type FindPetByIdResult = Pet;

export type FindPetByIdFallback = Error;

export interface DeletePetRequest {
  id: number
}
export type DeletePetResult = any
export type DeletePetFallback = Error;
"
`;

exports[`GenerateTypings given the Petstore schema should match generated snapshot 1`] = `
"export interface Pet {
  id: number;
  name: string;
  tag?: string;
  [k: string]: any;
}

export type Pets = (Pet)[];

export interface Error {
  code: number;
  message: string;
  [k: string]: any;
}

export interface ListPetsRequest {
  limit: number
}
export type ListPetsResult = Pets;

export type ListPetsFallback = Error;

export interface CreatePetsRequest {
}
export type CreatePetsResult = any
export type CreatePetsFallback = Error;

export interface ShowPetByIdRequest {
  petId: string
}
export type ShowPetByIdResult = Pets;

export type ShowPetById204 = any
export type ShowPetByIdFallback = Error;
"
`;
